File: ride-cleansing/src/main/scala/org/apache/flink/training/exercises/ridecleansing/scala/FlatMapExercise.scala
package org.apache.flink.training.exercises.ridecleansing.scala

import org.apache.flink.api.common.functions.{FilterFunction, FlatMapFunction}
import org.apache.flink.streaming.api.datastream.DataStream
import org.apache.flink.streaming.api.scala.{StreamExecutionEnvironment, createTypeInformation}
import org.apache.flink.training.exercises.common.EnrichedRide
import org.apache.flink.training.exercises.common.datatypes.TaxiRide
import org.apache.flink.util.Collector

object FlatMapExercise {

  def main(args: Array[String]): Unit = {
    val job = new FlatMapExerciseJob()

    job.runJob()
  }

  private class FlatMapExerciseJob {
    def runJob(): Unit = {

      val env = StreamExecutionEnvironment.getExecutionEnvironment

      val rideEvents = env.fromCollection(sourceRides())

      rideEvents.flatMap(new OnlyNYRidesEnriched).print()

      env.execute("JOSEMY - FlatMap exercise SCALA")
    }
    private def sourceRides(): List[TaxiRide] = {
      List[TaxiRide](
        new TaxiRide(1L, true),
        new TaxiRide(2L, false),
        new TaxiRide(3L, false),
        new TaxiRide(4L, true)
      )
    }
  }

  private class OnlyNYRidesEnriched() extends FlatMapFunction[TaxiRide, EnrichedRide] {
    override def flatMap(value: TaxiRide, out: Collector[EnrichedRide]): Unit = {
      val rideValidator = new RideCleansingExercise.NYCFilter()
      if (rideValidator.filter(value)) {
        out.collect(new EnrichedRide(value))
      }
    }
  }
}

File: ride-cleansing/src/main/java/org/apache/flink/training/exercises/ridecleansing/FlatMapExercise.java
package org.apache.flink.training.exercises.ridecleansing;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.training.exercises.common.EnrichedRide;
import org.apache.flink.training.exercises.common.datatypes.TaxiRide;
import org.apache.flink.util.Collector;

import java.util.ArrayList;
import java.util.List;

public class FlatMapExercise {

    public static void main(String[] args) throws Exception {
        FlatMapExercise job = new FlatMapExercise();

        job.runJob();
    }

    public void runJob() throws Exception {

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        DataStream<TaxiRide> rideEvents = env.fromCollection(sourceRides());

        DataStream<EnrichedRide> nyEnrichedRides = rideEvents.flatMap(new OnlyNYRidesEnritched());

        nyEnrichedRides.print();

        env.execute("JOSEMY - FlatMap exercise");
    }

    /**
     * Return only one element.
     */
    public static class OnlyNYRidesEnritched implements FlatMapFunction<TaxiRide, EnrichedRide> {
        @Override
        public void flatMap(TaxiRide taxiRide, Collector<EnrichedRide> out) throws Exception {
            //Create function to filter by NY Rides
            FilterFunction<TaxiRide> rideValidator = new RideCleansingExercise.NYCFilter();
            if (rideValidator.filter(taxiRide)) {
                // If the Ride is in NY we collect it
                out.collect(new EnrichedRide(taxiRide));
            }
        }
    }

    private List<TaxiRide> sourceRides() {

        List<TaxiRide> sourceRides = new ArrayList<>();

        sourceRides.add(new TaxiRide(1L, true));
        sourceRides.add(new TaxiRide(2L, false));
        sourceRides.add(new TaxiRide(3L, false));
        sourceRides.add(new TaxiRide(4L, true));

        return sourceRides;
    }
}
